operator/controllers/helmrelease_controller.go

```diff
@@ -44,19 +68,34 @@ func (r *HelmReleaseReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error)
 	if err := r.Get(ctx, req.NamespacedName, &helmRelease); err != nil {
 		if apierrs.IsNotFound(err) {
 			log.Info("HelmRelease doesn't exist")
-			return ctrl.Result{}, nil
+			return ctrl.Result{}, r.onDelete(ctx, req.NamespacedName)
 		}
 
 		return ctrl.Result{}, err
 	}
 
-	log.Info("Make things happen", "test", helmRelease.Spec)
-
-	return ctrl.Result{}, nil
+	return ctrl.Result{}, r.onUpdate(ctx, helmRelease)
 }
 
 func (r *HelmReleaseReconciler) SetupWithManager(mgr ctrl.Manager) error {
 	return ctrl.NewControllerManagedBy(mgr).
 		For(&shipitv1beta1.HelmRelease{}).
 		Complete(r)
 }
+
+func (r *HelmReleaseReconciler) onDelete(ctx context.Context, name types.NamespacedName) error {
+	// Update HelmRelease Status to 'PENDING_DELETE' -- but the HelmRelease doesn't exist anymore :(
+	// Delete the release with helm
+	return errNotImplemented
+}
+
+func (r *HelmReleaseReconciler) onUpdate(ctx context.Context, rls shipitv1beta1.HelmRelease) error {
+	// Update HelmRelease Status to 'PENDING_INSTALL' or 'PENDING_UPGRADE'
```

Comment:
It seems like it's a best practice to do sequences of state changes and reconciliation asynchronously over multiple reconciliation events, see `5. Use asynchronous sync loops` [0] and [1] (grep "A very common pattern is to observe events on the resources that we create")

There's some mechanism for managing ownership of cluster objects, so that a reconciliation event for a HelmRelease is queued when a resource it owns is updated [2]. I'm not sure how this works with Helm managing the releases resources, I'm going to take a look at other operators using helm.

[0] https://cloud.google.com/blog/products/containers-kubernetes/best-practices-for-building-kubernetes-operators-and-stateful-apps
[1] https://blog.openshift.com/kubernetes-operators-best-practices/ 
[2] https://godoc.org/sigs.k8s.io/controller-runtime/pkg/controller/controllerutil#SetControllerReference